kotlinx-coroutines-core/common/src/channels/ConflatedBufferedChannel.kt

```diff
@@ -0,0 +1,177 @@
+    onUndeliveredElement: OnUndeliveredElement<E>? = null
+) : BufferedChannel<E>(capacity = capacity, onUndeliveredElement = onUndeliveredElement) {
+    private val lock = reentrantLock() // guards all channel operations
+
+    init {
+        require(onBufferOverflow !== SUSPEND) {
+            "This implementation does not support suspension for senders, use ${BufferedChannel::class.simpleName} instead"
+        }
+        require(capacity >= 1) {
+            "Buffered channel capacity must be at least 1, but $capacity was specified"
+        }
+    }
+
+    override suspend fun receive(): E {
+        // Acquire the lock in the beginning of receive operation.
+        // Once the synchronization is completed (either the first
+        // element is retrieved, the operation suspends, or this
+        // channel is discovered in the closed state without element),
+        // `onReceiveSynchronizationCompletion` must be called by the
+        // underneath buffered channel implementation.
+        lock.lock()
+        return super.receive()
+    }
+
+    override suspend fun receiveCatching(): ChannelResult<E> {
+        // Acquire the lock in the beginning of receive operation.
+        // It automatically releases once the synchronization completes.
+        // See `receive` operation for details.
+        lock.lock()
+        return super.receiveCatching()
+    }
+
+    override fun tryReceive(): ChannelResult<E> {
+        // Acquire the lock in the beginning of receive operation.
+        // It automatically releases once the synchronization completes.
+        // See `receive` operation for details.
+        lock.lock()
+        return super.tryReceive()
+    }
+
+    override fun registerSelectForReceive(select: SelectInstance<*>, ignoredParam: Any?) {
+        // Acquire the lock in the beginning of receive operation.
+        // It automatically releases once the synchronization completes.
+        // See `receive` operation for details.
+        lock.lock()
+        super.registerSelectForReceive(select, ignoredParam)
+    }
+
+    override fun iterator(): ChannelIterator<E> = ConflatedChannelIterator()
+
+    private inner class ConflatedChannelIterator : BufferedChannelIterator() {
+        override suspend fun hasNext(): Boolean {
+            // Acquire the lock in the beginning of receive operation.
+            // It automatically releases once the synchronization completes.
+            // See `receive` operation for details.
+            lock.lock()
+            return super.hasNext()
+        }
+    }
+
+    override fun onReceiveSynchronizationCompleted() {
+        // Release the lock once the receive synchronization is done.
+        lock.unlock()
+    }
+
+    override suspend fun send(element: E) {
+        // Should never suspend, implement via `trySend(..)`.
+        val attempt = trySend(element)
+        if (attempt.isClosed) {
+            onUndeliveredElement?.callUndeliveredElement(element, coroutineContext)
```

Comment:
I do not understand when to call `onUndeliveredElement` with `coroutineContext` and when without it. My intuition was "call it this way when possible", but I see that here we break the previous behavior. Could you please clarify?

Note: the same issue should appear in the `BufferedChannel` implementation.

Reply:
The intuition is simple -- an exception is rethrowed (aka "no coroutineContext") when there is a channel-specific caller and is reported to the global exception handler otherwise.

Example of the former:

* When throwing `oUE` is invoked from `send`/`receive` that observed cancelled waiter  
* When throwing `oUE` is invoked from `close`/`cancel`, all exceptions are aggregated and thrown from the corresponding `close`/`cancel` call

Examples of the latter:

* When throwing `oUE` is invoked from an arbitrary place of code. Example: the receiver has been dispatched with an element, somebody called `cancel` on the `Job` and the receiver is cancelled in-flight. In such scenario there is no reasonable place the cancellation can be rethrown.

Also, note that new select implementation might have violated this contract, consider it a bug

Reply:
In Slack, we've decided that you will fix the issue to become more familiar with the code base.